pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -465,18 +476,41 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 
 // runNextSchedulableTask gets the next schedulable Tasks from the dag based on the current
 // pipeline run state, and starts them
-func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
+// after all DAG tasks are done, its responsible for scheduling final tasks and start executing them
+func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, dfinally *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
 
 	logger := logging.FromContext(ctx)
 	recorder := controller.GetEventRecorder(ctx)
 
-	candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulPipelineTaskNames()...)
-	if err != nil {
-		logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
-		return nil
+	var nextRprts []*resources.ResolvedPipelineRunTask
+	// When the pipeline run is stopping, we don't schedule any new task and only
+	// wait for all running tasks to complete and report their status
+	// after running tasks are done, check if there are any final task which needs to be scheduled
+	if pipelineState.IsStopping(d) {
```

Comment:
say we have task A and B running in parallel, with C as a finally task - if Task A fails, will IsStopping return true? i would think we would want to wait until "is stopped" or similar to run the finally task, not "is stopping"? (might just be the function name throwing me off)

Reply:
@bobcatfish see my comments in https://github.com/tektoncd/pipeline/pull/2661#pullrequestreview-430127441 and https://github.com/tektoncd/pipeline/pull/2661#issuecomment-644068201 and @pritidesai replies to them.

The name of the function is correct, and is turns true in the case you mentioned. 
However `GetFinalTasks` will return nothing unitl `B` completes too.

The code works - @pritidesai added a reconciler test for this use case specifically - but I think the division of responsibilities between the `reconciler` module and `resources` module is a bit mixed up. 
This was already the case though, before this PR, so my proposal here was to:
- keep the logic as proposed here in this PR
- solve the problem in https://github.com/tektoncd/pipeline/pull/2821 

Does it make sense? wdyt?

Reply:
I like the idea of having the reconciler call one function to get the next tasks to execute, and having that function look at the state of the pipelinerun, and return the next tasks to execute.

> keep the logic as proposed here in this PR, solve the problem in #2821

@afrittoli if we merged #2821, would `GetNextTasks` be responsible for looking at the tasks and returning either DAG tasks and finally tasks as appropriate?

If so I like that plan but I'd like it if there is something we could do in the meantime in this PR to make this confusing - i think all it would take to make that the case would be to call something other than "is stopping" that could tell that all the non-finally tasks failed, which would mean we want to execute finally tasks if they exist, and have GetFinalTasks always return the final tasks (only being called when they are actually needed)

Reply:
To simplify changes here, we could certainly replace what we have:

```
if pipelineState.IsStopping(d) {
  nextRprts = pipelineState.GetFinalTasks(d, dfinally)
} else {
  candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulOrSkippedDAGTasks(d)...)
  nextRprts = pipelineState.GetNextTasks(candidateTasks)
  if len(nextRprts) == 0 {
    nextRprts = pipelineState.GetFinalTasks(d, dfinally)
  }
}
```

To:

Option 1:

```
if !pipelineState.IsStopping(d) {
  candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulOrSkippedDAGTasks(d)...)
  nextRprts = pipelineState.GetNextTasks(candidateTasks)
}

nextRprts = append(nextRprts, pipelineState.GetFinalTasks(d, dfinally)...)
```
Pros/Cons:
- `GetFinalTasks` is called irrespective of what state the pipeline is in but `GetFinalTasks` is designed such that it wouldnt return anything unless its time to execute by looking at the pipeline state 
+ simpler than before 

Option 2:

```
if !pipelineState.IsStopping(d) {
  candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulOrSkippedDAGTasks(d)...)
  nextRprts = pipelineState.GetNextTasks(candidateTasks)
}

if len(nextRprts) == 0 {
  nextRprts = pipelineState.GetFinalTasks(d, dfinally)
}
```

Pros/Cons:
+ little simpler than what we have and Option 1
- Can create a false impression while reading the code that final tasks are executed when `nextRprts` is empty (similar to what we have) 😜

Reply:
sorry to speak up on behalf of @afrittoli, yes `GetNextTasks()` will be responsible for returning DAG/final/no task at all.

Reply:
Option 3:

```
nextRprts := pipelineState.GetExecutionQueue(d, dfinally)
```

`GetExecutionQueue` calls `GetNextTasks` and `GetFinalTasks` depending on the pipeline state.

```
func (state PipelineRunState) GetExecutionQueue(d *dag.Graph, dfinally *dag.Graph) []*ResolvedPipelineRunTask {
	var nextRprts []*ResolvedPipelineRunTask
	if !state.IsStopping(d) {
		candidateTasks, err := dag.GetSchedulable(d, state.SuccessfulOrSkippedDAGTasks(d)...)
		if err != nil {
			logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
			return nil
		}
		nextRprts = state.GetNextTasks(candidateTasks)
    }		
    if len(nextRprts) == 0 {
			nextRprts = state.GetFinalTasks(d, dfinally)
		}
	}
	return nextRprts
}
```

Reply:
adapted option 1 👍